### PR TITLE
Add PLUMgrid plugin support.

### PR DESCRIPTION
**Pull request by salmankh**
Define the resources to configure PLUMgrid plugin with Neutron.

Signed-off-by: salmank <salmank@plumgrid.com>

### CODE REVIEW

File: lib/puppet/parser/functions/device_for_network.rb [hunk 1]
```diff
@@ -12,7 +12,7 @@ module Puppet::Parser::Functions
 
     range = IPAddr.new(args[0])
     facts = compiler.node.facts.values
-    ip_addresses = facts.select { |key, value| key.match /^ipaddress_(.*)/ }
+    ip_addresses = facts.select { |key, value| key =~ /^ipaddress/ }
```
Comment (cmurphy) [inline lib/puppet/parser/functions/device_for_network.rb]: Why was this change made?

File: lib/puppet/parser/functions/ip_for_network.rb [hunk 2]
```diff
@@ -12,7 +12,7 @@ module Puppet::Parser::Functions
 
     range = IPAddr.new(args[0])
     facts = compiler.node.facts.values
-    ip_addresses = facts.select { |key, value| key.match /^ipaddress/ }
+    ip_addresses = facts.select { |key, value| key =~ /^ipaddress/ }
```
Comment (cmurphy) [inline lib/puppet/parser/functions/ip_for_network.rb]: Why was this change made?

File: manifests/common/plumgrid.pp [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# This class installs and configures Plumgrid Neutron Plugin.
+
+class openstack::common::plumgrid {
+  $is_controller = $::openstack::profile::base::is_controller
+  $controller_management_address = $::openstack::config::controller_address_management
+  $user = $::openstack::config::mysql_user_neutron
+  $pass = $::openstack::config::mysql_pass_neutron
+  $db_connection = "mysql://${user}:${pass}@${controller_management_address}/neutron"
+
+  if $is_controller {
```
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: Instead of using if $is_controller, this whole section could just be moved to the neutron server profile, leaving only the part that is actually common to multiple profiles in the common::plumgrid class.

File: manifests/common/plumgrid.pp [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+    file_line { '/etc/libvirt/qemu.conf group':
+      path   => '/etc/libvirt/qemu.conf',
+      line   => 'group = "root"',
+      match  => 'group =',
+      notify => Service['libvirt'],
+    }
+
+    Package['libvirt'] -> File_line<| path == '/etc/libvirt/qemu.conf' |>
+  }
+}
```
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: Looking at the documentation in https://wiki.openstack.org/wiki/PLUMgrid-Neutron I don't see anything that mentions changing qemu.conf. This seems to be out of scope for this change. Can you point me to some documentation that says otherwise?

File: manifests/profile/neutron/agent.pp [hunk 5]
```diff
@@ -1,5 +1,11 @@
 # The profile to set up a neutron agent
 class openstack::profile::neutron::agent {
   include ::openstack::common::neutron
-  include ::openstack::common::ml2::ovs
+
+  if ($::openstack::config::neutron_core_plugin != "plumgrid") {
```
Comment (cmurphy) [inline manifests/profile/neutron/agent.pp]: The logic is a little backward here, would rather see a case statement here selecting on neutron_core_plugin so that it's easier to add more plugins later.

File: lib/puppet/parser/functions/device_for_network.rb (hunk 1 above)
Comment (salmankh) [inline lib/puppet/parser/functions/device_for_network.rb]: This is to support the lower version of puppet. I was using puppet-3.1.1-1.el6 which results in failure due to non-support of key.match fucntion.

File: lib/puppet/parser/functions/ip_for_network.rb (hunk 2 above)
Comment (salmankh) [inline lib/puppet/parser/functions/ip_for_network.rb]: This is to support the lower version of puppet. I was using puppet-3.1.1-1.el6 which results in failure due to non-support of key.match fucntion.

File: manifests/common/plumgrid.pp (hunk 4 above)
Comment (salmankh) [inline manifests/common/plumgrid.pp]: Yah, will remove this. not needed.

File: manifests/profile/neutron/agent.pp (hunk 5 above)
Comment (salmankh) [inline manifests/profile/neutron/agent.pp]: ok. I will add this.

File: manifests/common/plumgrid.pp (hunk 3 above)
Comment (salmankh) [inline manifests/common/plumgrid.pp]: ok, will do it.

**Comment by salmankh** [comment]
@cmurphy Can you please review again. I have addressed your comments.

File: manifests/common/plumgrid.pp [hunk 6]
```diff
@@ -0,0 +1,16 @@
...
+  sysctl::value { 'net.ipv4.ip_forward': value => '1' }
+
+  # ifc_ctl_pp needs to be invoked by root as part of the vif.py when a VM is powered on
+  file { '/etc/sudoers.d/ifc_ctl_sudoers':
+    ensure  => file,
+    owner   => root,
+    group   => root,
+    mode    => '0440',
+    content => 'nova ALL=(root) NOPASSWD: /opt/pg/bin/ifc_ctl_pp *\n',
+  }
```
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: I can't find any documentation that says the nova user needs sudo for plumgrid or anything about ifc_ctl_pp, can you explain?

File: manifests/common/plumgrid.pp [hunk 7]
```diff
@@ -0,0 +1,16 @@
+# Common class used by Plumgrid Neutron Plugin.
+
+class openstack::common::plumgrid {
+  # forward all ipv4 traffic
+  # this is required for the vms to pass through the gateways public interface
+  sysctl::value { 'net.ipv4.ip_forward': value => '1' }
```
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: This resource is already declared https://github.com/puppetlabs/puppetlabs-openstack/blob/master/manifests/profile/neutron/router.pp#L3 so this will cause a duplicate declaration error.

File: manifests/common/plumgrid.pp (hunk 6 above)
Comment (salmankh) [inline manifests/common/plumgrid.pp]: ifc_ctl and ifc_ctl_pp are command line tools that provide the equivalent to 'brctl addif' for iovisor. It is used from nova's vif.py [https://github.com/openstack/nova/blob/master/nova/virt/libvirt/vif.py] in the PLUMgrid specific section.

File: manifests/common/plumgrid.pp (hunk 7 above)
Comment (salmankh) [inline manifests/common/plumgrid.pp]: Router profile(profile/neutron/router.pp) should be excluded for PLUMgrid deployments, it's not needed.

File: manifests/common/plumgrid.pp (hunk 6 above)
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: Are there docs supporting this configuration?

Comment (salmankh) [inline manifests/common/plumgrid.pp]: https://wiki.openstack.org/wiki/PLUMgrid-Neutron

File: manifests/common/plumgrid.pp (hunk 7 above)
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: Okay

File: manifests/common/plumgrid.pp (hunk 6 above)
Comment (cmurphy) [inline manifests/common/plumgrid.pp]: Okay

File: manifests/profile/neutron/server.pp [hunk 8]
```diff
@@ -12,11 +12,40 @@
   $tunnel_id_ranges              = $::openstack::config::neutron_tunnel_id_ranges # ['1:1000']
   $controller_management_address = $::openstack::config::controller_address_management
 
-  class  { '::neutron::plugins::ml2':
-    type_drivers         => $type_drivers,
-    tenant_network_types => $tenant_network_type,
-    mechanism_drivers    => $mechanism_drivers,
-    tunnel_id_ranges     => $tunnel_id_ranges
+  if ($::openstack::config::neutron_core_plugin == 'ml2') {
```
Comment (cmurphy) [inline manifests/profile/neutron/server.pp]: The ml2 core plugin is called 'neutron.plugins.ml2.plugin.Ml2Plugin', not simply 'ml2' https://github.com/puppetlabs/puppetlabs-openstack/blob/master/manifests/config.pp#L57

File: manifests/profile/neutron/agent.pp [hunk 9]
```diff
@@ -1,5 +1,10 @@
 # The profile to set up a neutron agent
 class openstack::profile::neutron::agent {
   include ::openstack::common::neutron
-  include ::openstack::common::ml2::ovs
+
+  case $::openstack::config::neutron_core_plugin {
+    'plumgrid': { include ::openstack::common::plumgrid }
```
Comment (cmurphy) [inline manifests/profile/neutron/agent.pp]: The core plugin that this parameter refers to is for a setting in neutron.conf and it looks like it is supposed to be 'neutron.plugins.plumgrid.plumgrid_plugin.plumgrid_plugin.NeutronPluginPLUMgridV2', not simply 'plumgrid'

File: manifests/config.pp [hunk 10]
```diff
@@ -68,6 +68,9 @@
   $neutron_type_drivers = undef,
   $neutron_mechanism_drivers = undef,
   $neutron_tunnel_id_ranges = undef,
+  $plumgrid_director_vip= undef,
```
Comment (bmjen) [inline manifests/config.pp]: Whitespacing needs fixing here.

File: manifests/init.pp [hunk 11]
```diff
@@ -416,6 +425,9 @@
   $neutron_shared_secret = undef,
   $neutron_core_plugin = undef,
   $neutron_service_plugins = undef,
+  $plumgrid_director_vip= undef,
```
Comment (bmjen) [inline manifests/init.pp]: white spacing again.

File: manifests/profile/neutron/server.pp (hunk 8 above)
Comment (salmankh) [inline manifests/profile/neutron/server.pp]: If you see the example at https://github.com/puppetlabs/puppetlabs-openstack/blob/master/examples/allinone.yaml,  its been overriden by "ml2". 
Also at https://github.com/openstack/neutron/blob/master/etc/neutron.conf entrypoint names of the plugins is supported instead the whole class name of a plugin be specified.

File: manifests/profile/neutron/agent.pp (hunk 9 above)
Comment (salmankh) [inline manifests/profile/neutron/agent.pp]: Same here using neutron entrypoint names of the plugins instead the whole class name.

File: manifests/profile/neutron/server.pp (hunk 8 above)
Comment (salmankh) [inline manifests/profile/neutron/server.pp]: @cmurphy Are you okay with plugins entry point names used here or you still want me to use the whole class name?

Comment (cmurphy) [inline manifests/profile/neutron/server.pp]: You're right, this seems fine.

**Comment by salmankh** [comment]
Thanks @cmurphy for your deep reiew of things. I believe its good to merge now, not sure who can do it.
